lib/grandpa/finalization.go

```
@@ -0,0 +1,493 @@
+type handleVotingRound struct {
+	grandpaService       *Service
+	finalizationEngineCh <-chan engineAction
+	stopCh               chan struct{}
+	engineDone           chan struct{}
+}
+
+func newHandleVotingRound(service *Service, finalizationEngineCh <-chan engineAction) *handleVotingRound {
+	return &handleVotingRound{
+		grandpaService:       service,
+		stopCh:               make(chan struct{}),
+		engineDone:           make(chan struct{}),
+		finalizationEngineCh: finalizationEngineCh,
+	}
+}
+
+func (h *handleVotingRound) Stop() (err error) {
+	if h.engineDone == nil {
+		return nil
+	}
+
+	close(h.stopCh)
+	<-h.engineDone
+
+	return nil
+}
+
+func (h *handleVotingRound) Run() (err error) {
+	defer func() {
+		close(h.engineDone)
+		h.engineDone = nil
+	}()
+
+	start := time.Now()
+
+	logger.Debugf("starting round %d with set id %d",
+		h.grandpaService.state.round, h.grandpaService.state.setID)
+
+	for {
+		select {
+		case <-h.stopCh:
+			return nil
+
+		case action, ok := <-h.finalizationEngineCh:
+			if !ok {
+				return nil
+			}
+
+			switch action {
+			case determinePrevote:
+				isPrimary, err := h.grandpaService.handleIsPrimary()
+				if err != nil {
+					return fmt.Errorf("handling primary: %w", err)
+				}
+
+				// broadcast pre-vote
+				preVote, err := h.grandpaService.determinePreVote()
+				if err != nil {
+					return fmt.Errorf("determining pre-vote: %w", err)
+				}
+
+				signedpreVote, prevoteMessage, err :=
+					h.grandpaService.createSignedVoteAndVoteMessage(preVote, prevote)
+				if err != nil {
+					return fmt.Errorf("creating signed vote: %w", err)
+				}
+
+				if !isPrimary {
+					h.grandpaService.prevotes.Store(h.grandpaService.publicKeyBytes(), signedpreVote)
+				}
+
+				logger.Debugf("sending pre-vote message: {%v}", prevoteMessage)
+				err = h.grandpaService.sendPrevoteMessage(prevoteMessage)
+				if err != nil {
+					logger.Errorf("sending pre-vote message: %s", err)
```

why not return an error instead?